kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,442 @@
+//! Defines [`LogReplayScanner`] used by [`TableChangesScan`] to process commit files and extract
+//! the metadata needed to generate the Change Data Feed.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, LazyLock};
+
+use crate::actions::visitors::{visit_deletion_vector_at, ProtocolVisitor};
+use crate::actions::{
+    get_log_add_schema, get_log_schema, Protocol, ADD_NAME, CDC_NAME, METADATA_NAME, PROTOCOL_NAME,
+    REMOVE_NAME,
+};
+use crate::engine_data::{GetData, TypedGetData};
+use crate::expressions::{column_expr, column_name, ColumnName, Expression};
+use crate::path::ParsedLogPath;
+use crate::scan::data_skipping::DataSkippingFilter;
+use crate::scan::scan_row_schema;
+use crate::scan::state::DvInfo;
+use crate::schema::{ArrayType, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructType};
+use crate::table_changes::{check_cdf_table_properties, ensure_cdf_read_supported};
+use crate::table_properties::TableProperties;
+use crate::utils::require;
+use crate::{DeltaResult, Engine, EngineData, Error, ExpressionRef, RowVisitor};
+use itertools::Itertools;
+
+#[cfg(test)]
+mod tests;
+
+/// Scan data for a Change Data Feed query. This holds metadata that is needed to read data rows.
+#[allow(unused)]
+pub(crate) struct TableChangesScanData {
+    /// Engine data with the schema defined in [`scan_row_schema`]
+    ///
+    /// Note: The schema of the engine data will be updated in the future columns used by
+    /// Change Data Feed.
```

Comment:
is this more accurate?
```suggestion
    /// Note: The schema of the engine data will be updated in the future to include columns
    /// used by Change Data Feed.
```

Reply:
Yep 👍